File: xiaohui/2.数据结构基础/LinkedList.java

public class LinkedList {
    private Node head;
    private Node last;
    private int size;

    private static class Node {
        int data;
        Node next;
        Node(int data) {
            this.data = data;
        }
    }

    public Node getNode(int index) throws Exception {
        if (index<0 || index>=size-1) {
            throw new IndexOutOfBoundsException();
        }
        Node temp = head;
        for (int i=0; i<=index; i++) {
            temp = temp.next;
        }
        return temp;
    }

    public void insert(int data, int index) throws Exception{
        if (index<0 || index>size) {
            throw new IndexOutOfBoundsException();
        }
        Node insertedNode = new Node(data);
        if (size == 0) {
            head = insertedNode;
            insertedNode.next = null;
        } else if (index == 0) {
            insertedNode.next = head;
            head = insertedNode;
        } else if (index == size) {
            last.next = insertedNode;
            insertedNode.next = null;
        } else {
            Node prevNode = getNode(index-1);
            insertedNode.next = prevNode.next;
            prevNode.next = insertedNode;
        }
        size++;
    }

    public Node remove(int index) throws Exception {
        if (index<0 || index>size) {
            throw new IndexOutOfBoundsException();
        }
        Node removedNode = null;
        if (index == 0) {
            removedNode = head;
            head = head.next;
        } else if (index == size-1) {
            Node prevNode = getNode(index-1);
            removedNode = last;
            last = prevNode;
            prevNode.next = null;
        } else {
            Node prevNode = getNode(index-1);
            removedNode = prevNode.next;
            prevNode.next = removedNode.next;
        }
        size--;
        return removedNode;
    }

    public void output() {
        Node temp = head;
        while (temp != null) {
            System.out.println(temp.data);
            temp = temp.next;
        }
    }
}

File: xiaohui/2.数据结构基础/LinkedList.py
class LinkedList:
    class Node:
        def __init__(self, data, next=None):
            self.data = data
            self.next = next

    def __init__(self, head, last):
        self.head = head
        self.last = last
        self.size = 0
    
    def getNode(self, index):
        if index<0 or index >= self.size-1:
            raise Exception("index is invalid")
        temp = self.head
        for i in range(index+1)
            temp = temp.next
        return temp
    
    def insert(data, index):
        if index<0 or index >= self.size-1:
            raise Exception("index is invalid")
        temp = Node(data)
        if size == 0:
            self.head = temp
            temp.next = None
        elif index == 0 
            temp.next = head
            head = temp
        elif index == size:
            last.next = temp
            temp.next = None
        else:
            prevNode = self.getNode(index-1)
            temp.next = prevNode.next
            prevNode.next = temp
        self.size += 1

    def remove(self, index):
        if index<0 or index >= self.size-1:
            raise Exception("index is invalid")
        temp = Node(data)
        if index == 0:
            temp = self.head
            self.head = self.head.next
        elif index == self,size-1
            temp = last
            prevNode = self.getNode(index-1)
            last = prevNode
            prevNode.next = None
        else:
            prevNode = self.getNode(index-1)
            temp = prevNode.next
            prevNode.next = prevNode.next.next
        self.size -= 1
        return temp

    def output(self):
        temp = self.head
        while temp.next is not None:
            print(temp.data)
            temp = temp.next
